server/src/internalClusterTest/java/org/opensearch/cluster/coordination/AwarenessAttributeDecommissionIT.java

```diff
@@ -56,6 +72,425 @@ public void cleanup() throws Exception {
+        });
+    }
+
+    public void testDecommissionFailedWhenNotForceZoneAware() throws Exception {
+        Settings commonSettings = Settings.builder().put("cluster.routing.allocation.awareness.attributes", "zone").build();
+        // Start 3 cluster manager eligible nodes
+        logger.info("--> start 3 cluster manager nodes on zones 'a' & 'b' & 'c'");
+        internalCluster().startNodes(
+            Settings.builder()
+                .put(commonSettings)
+                .put("node.attr.zone", "a")
+                .put(onlyRole(commonSettings, DiscoveryNodeRole.CLUSTER_MANAGER_ROLE))
+                .build(),
+            Settings.builder()
+                .put(commonSettings)
+                .put("node.attr.zone", "b")
+                .put(onlyRole(commonSettings, DiscoveryNodeRole.CLUSTER_MANAGER_ROLE))
+                .build(),
+            Settings.builder()
+                .put(commonSettings)
+                .put("node.attr.zone", "c")
+                .put(onlyRole(commonSettings, DiscoveryNodeRole.CLUSTER_MANAGER_ROLE))
+                .build()
+        );
+        logger.info("--> starting data node each on zones 'a' & 'b' & 'c'");
+        internalCluster().startDataOnlyNode(Settings.builder().put(commonSettings).put("node.attr.zone", "a").build());
+        internalCluster().startDataOnlyNode(Settings.builder().put(commonSettings).put("node.attr.zone", "b").build());
+        internalCluster().startDataOnlyNode(Settings.builder().put(commonSettings).put("node.attr.zone", "c").build());
+        ensureStableCluster(6);
+        ClusterHealthResponse health = client().admin()
+            .cluster()
+            .prepareHealth()
+            .setWaitForEvents(Priority.LANGUID)
+            .setWaitForGreenStatus()
+            .setWaitForNodes(Integer.toString(6))
+            .execute()
+            .actionGet();
+        assertFalse(health.isTimedOut());
+
+        DecommissionAttribute decommissionAttribute = new DecommissionAttribute("zone", "a");
+        DecommissionRequest decommissionRequest = new DecommissionRequest(decommissionAttribute);
+        assertBusy(() -> {
+            DecommissioningFailedException ex = expectThrows(
+                DecommissioningFailedException.class,
+                () -> client().execute(DecommissionAction.INSTANCE, decommissionRequest).actionGet()
+            );
+            assertTrue(ex.getMessage().contains("doesn't have the decommissioning attribute"));
+        });
+    }
+
+    public void testNodesRemovedAfterZoneDecommission_ClusterManagerNotInToBeDecommissionedZone() throws Exception {
+        assertNodesRemovedAfterZoneDecommission(false);
+    }
+
+    public void testNodesRemovedAfterZoneDecommission_ClusterManagerInToBeDecommissionedZone() throws Exception {
+        assertNodesRemovedAfterZoneDecommission(true);
+    }
+
+    public void testInvariantsAndLogsOnDecommissionedNodes() throws Exception {
```

Comment:
Methods are very big around 150 lines in some cases. It's hard to read in a single go. Can we do something about it?

Reply:
I tried to abstract out but use cases for each of these are unique and hence kept it like this